{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "95d24661",
   "metadata": {},
   "source": [
    "# Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7bc9629e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44808\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "import h5py\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import copy\n",
    "import tqdm\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import ast\n",
    "sys.path.append(r\"C:\\Users\\shiwei\\Documents\")\n",
    "\n",
    "print(os.getpid())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d9031d2",
   "metadata": {},
   "source": [
    "# Define functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb7b8c52",
   "metadata": {},
   "source": [
    "## functions to convert data format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fe89f190",
   "metadata": {},
   "outputs": [],
   "source": [
    "# funtion to convert\n",
    "def format_4dn_dataframe_trace (spot_df, pixel_size, required_cols,\n",
    "                                #ExpFovCell_2_uid  #old version\n",
    "                                 ):\n",
    "    \n",
    "    from tqdm import tqdm\n",
    "    print (f'Process dataframe with {len(spot_df)} spots.')\n",
    "    \n",
    "    # 1. get cell uid as CellID\n",
    "    ## old version that add uid now/ comment out\n",
    "    #df_4dn = spot_df[['FOV','orig_cellID']]   \n",
    "    #df_4dn['FOV_ID']= df_4dn['FOV'].apply(lambda x: str(x))\n",
    "    #df_4dn['CellID_byFOV'] = df_4dn['orig_cellID'].apply(lambda x: x.split('Cell-')[1])\n",
    "    #def ExpFovCell_2_uid_from_FovCell (Fov_id, Cell_id, ExpFovCell_2_uid, Exp_name='None'):\n",
    "        #if Exp_name=='None':\n",
    "            #Exp_name = list(ExpFovCell_2_uid.keys())[0][0]\n",
    "        #Cell_uid = ExpFovCell_2_uid[Exp_name,Fov_id, Cell_id]\n",
    "        #return Cell_uid\n",
    "        \n",
    "    #spot_FovCell_list = []\n",
    "    #for Fov_id, Cell_id in tqdm(zip(df_4dn['FOV'],df_4dn['orig_cellID'])):\n",
    "    #for Fov_id, Cell_id in zip(df_4dn['FOV_ID'],df_4dn['CellID_byFOV']):\n",
    "        #spot_FovCell_list.append(ExpFovCell_2_uid_from_FovCell (Fov_id, Cell_id, ExpFovCell_2_uid, Exp_name='None'))\n",
    "    #df_4dn['Cell_ID'] = df_4dn.apply(lambda x: ExpFovCell_2_uid_from_FovCell(x['FOV'],x['orig_cellID'],ExpFovCell_2_uid),axis=1)\n",
    "    #df_4dn['Cell_ID']=spot_FovCell_list\n",
    "    \n",
    "    ## new version that directly reads in uid from df\n",
    "    df_4dn = spot_df[['FOV','orig_cellID','finalcellID',]]\n",
    "    df_4dn['FOV_ID']= df_4dn['FOV'].apply(lambda x: str(x))\n",
    "    df_4dn['CellID_byFOV'] = df_4dn['orig_cellID'].apply(lambda x: x.split('Cell-')[1])\n",
    "    df_4dn['Cell_ID']= df_4dn['finalcellID'].apply(lambda x: str(x))\n",
    "    \n",
    "    \n",
    "    ##################################################\n",
    "    # 2. get other columns\n",
    "    other_cols = [_c for _c in required_cols if _c not in ['FOV','orig_cellID','finalcellID',]]\n",
    "    df_4dn[other_cols] = spot_df[other_cols]\n",
    "    df_4dn[['Z','X','Y']] = df_4dn[['z_hat','x_hat','y_hat']] * pixel_size / 1000\n",
    "    #df_4dn[['Chrom','Chrom_Start','Chrom_End','Chrom_order']] = df_4dn[['chr','start', 'end','hyb']]\n",
    "    df_4dn[['Chrom','Chrom_order']] = df_4dn[['chr','hyb']]\n",
    "    df_4dn['Chrom_Start'] = df_4dn['region_name'].apply(lambda x: x.split(':')[1].split('-')[0])\n",
    "    df_4dn['Chrom_End'] = df_4dn['region_name'].apply(lambda x: x.split(':')[1].split('-')[1])\n",
    "    \n",
    "    df_4dn['Trace_ID'] = df_4dn.apply(lambda x: '_'.join([x['FOV_ID'],\n",
    "                                                          x['CellID_byFOV'],\n",
    "                                                          x['Chrom'].split('chr')[1],\n",
    "                                                          str(x['fiberidx'])]),axis=1)\n",
    "    \n",
    "    df_4dn['Spot_ID'] = df_4dn['rowID']\n",
    "    \n",
    "    df_4dn.drop(columns=required_cols, inplace=True)\n",
    "    df_4dn=df_4dn[['Spot_ID','Trace_ID','X','Y','Z','Chrom','Chrom_Start','Chrom_End','Chrom_order','Cell_ID','FOV_ID','CellID_byFOV']]\n",
    "    \n",
    "    return df_4dn\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "import multiprocessing as mp\n",
    "from functools import partial\n",
    "import inspect\n",
    "\n",
    "# create temp py to parallel\n",
    "def parallel_task(func, zipped_iterables):\n",
    "    import os, time\n",
    "    cwd = os.getcwd()\n",
    "    print (\"Write in the function to multiprocess as a temp file.\")\n",
    "    \n",
    "    with open(os.path.join(cwd,f'tmp_func.py'), 'w') as file:\n",
    "        file.write(inspect.getsource(func).replace(func.__name__, \"function_to_mp\"))\n",
    "        \n",
    "    from tmp_func import function_to_mp\n",
    "\n",
    "    if __name__ == '__main__':\n",
    "        start_time = time.time()\n",
    "        with mp.Pool(16) as mp_pool:\n",
    "            mp_res = mp_pool.starmap(function_to_mp, zipped_iterables,chunksize=1)\n",
    "            mp_pool.close()\n",
    "            mp_pool.join()\n",
    "            mp_pool.terminate()\n",
    "        \n",
    "        elapsed_time = time.time() - start_time\n",
    "        hours, rem = divmod(elapsed_time, 3600)\n",
    "        minutes, seconds = divmod(rem, 60)\n",
    "        print (f\"Complete multiprocess; remove the temp file for the function.\")\n",
    "        print(\"{:0>2}:{:0>2}:{:05.2f}\".format(int(hours),int(minutes),seconds))\n",
    "        if os.path.exists(os.path.join(cwd,f'tmp_func.py')):\n",
    "            os.remove(os.path.join(cwd,f'tmp_func.py'))\n",
    "        \n",
    "        return mp_res\n",
    "    else:\n",
    "        raise \"Not in Jupyter Notebook\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78373733",
   "metadata": {},
   "source": [
    "## prepare header lines "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fa888b92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'4dn_FOF-CT_cell, 4dn_FOF-CT_rna_spot, 4dn_FOF-CT_demultiplexing'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_name_list = ['4dn_FOF-CT_core', '4dn_FOF-CT_cell', '4dn_FOF-CT_rna_spot', '4dn_FOF-CT_demultiplexing']\n",
    "curr_table_name = '4dn_FOF-CT_core'\n",
    "add_table_name = ', '.join([t for t in table_name_list if t!=curr_table_name])\n",
    "add_table_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "036c2d2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# change header line below manually\n",
    "header_dict_v0 = {'##FOF-CT_version=':'v0.1',\n",
    "                  '##Table_namespace=':curr_table_name,\n",
    "                  '#lab_name:':'Xiaowei Zhuang lab', \n",
    "                  '#experimenter_name:':'Pu Zheng, Shiwei Liu',\n",
    "                  '#experimenter_contact:':'pu_zheng@g.harvard.edu, shiweiliu@fas.harvard.edu',\n",
    "                  '#experiment_description:':'This experiment set contains integrated RNA- and DNA-MERFISH of the mouse MOp region and a portion of adjacent regions',\n",
    "                  '#table_description:':'This table contains picked DNA spots for DNA fibers in inidividual cells from the mouse MOp region and a portion of adjacent regions',\n",
    "                  '#Software_Title:':'ImageAnalysis3, jie',\n",
    "                  '#Software_Type:':'SpotLoc(by ImageAnalysis3), Tracing(by jie)',\n",
    "                  '#Software_Authors:':'Pu Zheng(ImageAnalysis3), Blair Bojing Jia(jie)',\n",
    "                  '#Software_Description:':'ImageAnalysis3 for decoding and partition, jie for picking',\n",
    "                  '#Software_Repository:':'https://github.com/zhengpuas47/ImageAnalysis3, https://github.com/b2jia/jie',\n",
    "                  '#Software_PreferredCitationID:':'https://doi: 10.1016/j.cell.2020.07.032, https://doi.org/10.1101/2022.03.25.485845',\n",
    "                  #'#additional_tables:':'4dn_FOF-CT_rna, 4dn_FOF-CT_quality, 4dn_FOF-CT_bio, 4dn_FOF-CT_demultiplexing, 4dn_FOF-CT_trace, 4dn_FOF-CT_cell',\n",
    "                  '#additional_tables:':add_table_name,\n",
    "                  '##genome_assembly=':'GRCm38(mm10)',\n",
    "                  '##XYZ_unit=':'nm',\n",
    "                  '##columns=':'(Spot_ID, Trace_ID, X, Y, Z, Chrom, Chrom_Start, Chrom_End, Chrom_order, Cell_ID, FOV_ID, CellID_byFOV)',\n",
    "                  '#Spot_ID:':'unique DNA spot identifier across 4dn_FOF-CT_core and 4dn_FOF-CT_demultiplxeing for the same replicate',\n",
    "                  '#Trace_ID:':'number format as fovXX_cellXX_chrXX_chromatidXX',\n",
    "                  '#Chrom_order:':'order of DNA region on its respective chromosome',\n",
    "                  '#Cell_ID:':'unique cell identifier across the datatables',\n",
    "                  '#FOV:':'FOV ID across datatables for the same replicate',\n",
    "                  '#CellID_byFOV':'specific cell identifier relative to the FOV',\n",
    "\n",
    "                \n",
    "                 }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8c576a77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'##FOF-CT_version=': 'v0.1',\n",
       " '##Table_namespace=': '4dn_FOF-CT_core',\n",
       " '#lab_name:': 'Xiaowei Zhuang lab',\n",
       " '#experimenter_name:': 'Pu Zheng, Shiwei Liu',\n",
       " '#experimenter_contact:': 'pu_zheng@g.harvard.edu, shiweiliu@fas.harvard.edu',\n",
       " '#experiment_description:': 'This experiment set contains integrated RNA- and DNA-MERFISH of the mouse MOp region and a portion of adjacent regions',\n",
       " '#table_description:': 'This table contains picked DNA spots for DNA fibers in inidividual cells from the mouse MOp region and a portion of adjacent regions',\n",
       " '#Software_Title:': 'ImageAnalysis3, jie',\n",
       " '#Software_Type:': 'SpotLoc(by ImageAnalysis3), Tracing(by jie)',\n",
       " '#Software_Authors:': 'Pu Zheng(ImageAnalysis3), Blair Bojing Jia(jie)',\n",
       " '#Software_Description:': 'ImageAnalysis3 for decoding and partition, jie for picking',\n",
       " '#Software_Repository:': 'https://github.com/zhengpuas47/ImageAnalysis3, https://github.com/b2jia/jie',\n",
       " '#Software_PreferredCitationID:': 'https://doi: 10.1016/j.cell.2020.07.032, https://doi.org/10.1101/2022.03.25.485845',\n",
       " '#additional_tables:': '4dn_FOF-CT_cell, 4dn_FOF-CT_rna_spot, 4dn_FOF-CT_demultiplexing',\n",
       " '##genome_assembly=': 'GRCm38(mm10)',\n",
       " '##XYZ_unit=': 'nm',\n",
       " '##columns=': '(Spot_ID, Trace_ID, X, Y, Z, Chrom, Chrom_Start, Chrom_End, Chrom_order, Cell_ID, FOV_ID, CellID_byFOV)',\n",
       " '#Spot_ID:': 'unique DNA spot identifier across 4dn_FOF-CT_core and 4dn_FOF-CT_demultiplxeing for the same replicate',\n",
       " '#Trace_ID:': 'number format as fovXX_cellXX_chrXX_chromatidXX',\n",
       " '#Chrom_order:': 'order of DNA region on its respective chromosome',\n",
       " '#Cell_ID:': 'unique cell identifier across the datatables',\n",
       " '#FOV:': 'FOV ID across datatables for the same replicate',\n",
       " '#CellID_byFOV': 'specific cell identifier relative to the FOV'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "header_dict_v0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "de7aa4ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>##FOF-CT_version=v0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>##Table_namespace=4dn_FOF-CT_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>#lab_name:Xiaowei Zhuang lab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>#experimenter_name:Pu Zheng, Shiwei Liu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>#experimenter_contact:pu_zheng@g.harvard.edu, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>#experiment_description:This experiment set co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>#table_description:This table contains picked ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>#Software_Title:ImageAnalysis3, jie</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>#Software_Type:SpotLoc(by ImageAnalysis3), Tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>#Software_Authors:Pu Zheng(ImageAnalysis3), Bl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>#Software_Description:ImageAnalysis3 for decod...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>#Software_Repository:https://github.com/zhengp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>#Software_PreferredCitationID:https://doi: 10....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>#additional_tables:4dn_FOF-CT_cell, 4dn_FOF-CT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>##genome_assembly=GRCm38(mm10)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>##XYZ_unit=nm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>##columns=(Spot_ID, Trace_ID, X, Y, Z, Chrom, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>#Spot_ID:unique DNA spot identifier across 4dn...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>#Trace_ID:number format as fovXX_cellXX_chrXX_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>#Chrom_order:order of DNA region on its respec...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>#Cell_ID:unique cell identifier across the dat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>#FOV:FOV ID across datatables for the same rep...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>#CellID_byFOVspecific cell identifier relative...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0                               ##FOF-CT_version=v0.1\n",
       "1                   ##Table_namespace=4dn_FOF-CT_core\n",
       "2                        #lab_name:Xiaowei Zhuang lab\n",
       "3             #experimenter_name:Pu Zheng, Shiwei Liu\n",
       "4   #experimenter_contact:pu_zheng@g.harvard.edu, ...\n",
       "5   #experiment_description:This experiment set co...\n",
       "6   #table_description:This table contains picked ...\n",
       "7                 #Software_Title:ImageAnalysis3, jie\n",
       "8   #Software_Type:SpotLoc(by ImageAnalysis3), Tra...\n",
       "9   #Software_Authors:Pu Zheng(ImageAnalysis3), Bl...\n",
       "10  #Software_Description:ImageAnalysis3 for decod...\n",
       "11  #Software_Repository:https://github.com/zhengp...\n",
       "12  #Software_PreferredCitationID:https://doi: 10....\n",
       "13  #additional_tables:4dn_FOF-CT_cell, 4dn_FOF-CT...\n",
       "14                     ##genome_assembly=GRCm38(mm10)\n",
       "15                                      ##XYZ_unit=nm\n",
       "16  ##columns=(Spot_ID, Trace_ID, X, Y, Z, Chrom, ...\n",
       "17  #Spot_ID:unique DNA spot identifier across 4dn...\n",
       "18  #Trace_ID:number format as fovXX_cellXX_chrXX_...\n",
       "19  #Chrom_order:order of DNA region on its respec...\n",
       "20  #Cell_ID:unique cell identifier across the dat...\n",
       "21  #FOV:FOV ID across datatables for the same rep...\n",
       "22  #CellID_byFOVspecific cell identifier relative..."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert header lines to dataframe\n",
    "header_lines = [k+v for k,v in header_dict_v0.items()]\n",
    "header_lines_df = pd.DataFrame(header_lines)\n",
    "header_lines_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66447322",
   "metadata": {},
   "source": [
    "# Load and convert spot dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6309130d",
   "metadata": {},
   "source": [
    "## define parameters for replicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a9ab83b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pick main folder; where sorted spots are saved\n",
    "pick_folder = r'\\\\10.245.74.158\\Chromatin_NAS_8\\Exported_data\\jie_CW_decode'\n",
    "# paramter version used for pick\n",
    "pick_param_name = 'CW_v2_sorted'\n",
    "\n",
    "# define name and other metadata info in correct order for each experiment below\n",
    "exp_name_list = ['20220316exp', '20220402exp', '20220419exp', '20220713exp']\n",
    "RNA_experiment_ID_list = ['20220304','20220329','20220415','20220418']\n",
    "DNA_experiment_ID_list = ['20220316','20220402','20220419','20220713']\n",
    "sample_ID_list = ['C57BL/6_M_1_MOp_1','C57BL/6_M_2_MOp_1','C57BL/6_M_3_MOp_1','C57BL/6_M_3_MOp_2']\n",
    "\n",
    "exp_sample_type = 'WT' # to further distinguish the output exp name in case the date is the same\n",
    "# naming for 4DN alias\n",
    "Bio_num_list = [1,2,3,3]\n",
    "Tech_num_list = [1,1,1,2]\n",
    "table_type = curr_table_name.split('4dn_FOF-CT_')[1]\n",
    "\n",
    "\n",
    "# output folder to save the converted df\n",
    "output_folder =r'F:\\4DN_deposit\\new_version'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4d5987d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# shared parameters\n",
    "# required cols from the data to be processed\n",
    "required_cols = ['FOV', 'chr', #'start', 'end',\n",
    "       'hyb', 'x_hat', 'y_hat', 'z_hat', \n",
    "        'finalcellID', 'region_name',\n",
    "       'orig_cellID','fiberidx','rowID']\n",
    "\n",
    "pixel_size = 108 #'z_hat','x_hat','y_hat' are pseduo pixels where z_hat unit is equvilant to x_hat "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94528901",
   "metadata": {},
   "source": [
    "## process data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "917fc393",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Load the pick result as 20220316exp_pick_chr1.h5 with 20 files in total.\n",
      "Start processing.\n",
      "Write in the function to multiprocess as a temp file.\n",
      "Complete multiprocess; remove the temp file for the function.\n",
      "00:00:31.04\n",
      "Result saved.\n",
      "Load the pick result as 20220402exp_pick_chr1.h5 with 20 files in total.\n",
      "Start processing.\n",
      "Write in the function to multiprocess as a temp file.\n",
      "Complete multiprocess; remove the temp file for the function.\n",
      "00:00:30.72\n",
      "Result saved.\n",
      "Load the pick result as 20220419exp_pick_chr1.h5 with 20 files in total.\n",
      "Start processing.\n",
      "Write in the function to multiprocess as a temp file.\n",
      "Complete multiprocess; remove the temp file for the function.\n",
      "00:00:29.86\n",
      "Result saved.\n",
      "Load the pick result as 20220713exp_pick_chr1.h5 with 20 files in total.\n",
      "Start processing.\n",
      "Write in the function to multiprocess as a temp file.\n",
      "Complete multiprocess; remove the temp file for the function.\n",
      "00:00:39.92\n",
      "Result saved.\n"
     ]
    }
   ],
   "source": [
    "# loop through all replicate and batch process each\n",
    "for _exp_ind, _exp_name in enumerate(exp_name_list):\n",
    "\n",
    "    pick_subfolder = os.path.join(pick_folder,_exp_name,pick_param_name)\n",
    "    \n",
    "    experiment_key= RNA_experiment_ID_list[_exp_ind] + '_' + exp_sample_type\n",
    "    #data_4dn_savename = os.path.join(output_folder, f'{experiment_key}_core_table.csv')\n",
    "    Bio_num = Bio_num_list[_exp_ind]\n",
    "    Tech_num = Tech_num_list[_exp_ind]\n",
    "    data_4dn_savename = os.path.join(output_folder, f'fileproc_genome_wide_brain_ct_B{Bio_num}_T{Tech_num}_{table_type}_f1.csv')\n",
    "    if os.path.exists(data_4dn_savename):\n",
    "        print (f'Output already exists for {experiment_key}. Skip.')\n",
    "        \n",
    "    else:\n",
    "        # output dict to load chromosome pick h5\n",
    "        output_df_dict = {}\n",
    "\n",
    "        if os.path.exists(pick_subfolder):\n",
    "            # 1. load data\n",
    "            file_list = [os.path.join(pick_subfolder, _fl) for _fl in os.listdir(pick_subfolder) if '.h5' in _fl]\n",
    "            for file in file_list:\n",
    "                output_df_dict[os.path.basename(file)]=pd.read_hdf(file)\n",
    "            print(f'Load the pick result as {list(output_df_dict.keys())[0]}', end=' ')\n",
    "            print(f'with {len(list(output_df_dict.keys()))} files in total.')\n",
    "\n",
    "            # 2. batch convert data\n",
    "            print(f'Start processing.')\n",
    "            spot_df_list = list(output_df_dict.values())\n",
    "            pixel_size_list = [pixel_size,]*len(spot_df_list)\n",
    "            required_cols_list = [required_cols,]*len(spot_df_list)\n",
    "\n",
    "            df_4dn_list = parallel_task(format_4dn_dataframe_trace, zip(spot_df_list,\n",
    "                                                          pixel_size_list,\n",
    "                                                          required_cols_list,\n",
    "                                                          #ExpFovCell_2_uid_list\n",
    "                                                               ))\n",
    "\n",
    "\n",
    "            df_4dn_all= pd.concat(df_4dn_list)\n",
    "            df_4dn_all['RNA_experiment_ID'] = RNA_experiment_ID_list[_exp_ind]\n",
    "            df_4dn_all['DNA_experiment_ID'] = DNA_experiment_ID_list[_exp_ind]\n",
    "            df_4dn_all['Sample_ID'] = sample_ID_list[_exp_ind]\n",
    "\n",
    "            # 3. reset index after pd concat\n",
    "            df_4dn_all.reset_index(inplace=True,drop=True)\n",
    "            df_4dn_all_sort = df_4dn_all.sort_values(by='Spot_ID')\n",
    "\n",
    "\n",
    "            # 4. save output\n",
    "            # modify column names and add 4dn header lines\n",
    "            header_lines_df_fill = header_lines_df.reindex(columns = list(np.arange(0,len(df_4dn_all_sort.columns))))\n",
    "            df_4dn_new = pd.DataFrame(np.vstack([df_4dn_all_sort.columns,df_4dn_all_sort]))\n",
    "            df_4dn_clean = pd.concat([header_lines_df_fill,df_4dn_new])\n",
    "\n",
    "            df_4dn_clean.to_csv(data_4dn_savename, header=None, index=False)\n",
    "            print(f'Result saved.')\n",
    "\n",
    "\n",
    "        else:\n",
    "            print ('Pick results not availale. Check results.')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
